sparse_strips/vello_hybrid/src/strip.rs

```diff
@@ -0,0 +1,209 @@
+
+    // Comparison function for sorting. Only compares loc, doesn't care
+    // about points. Unpacking code has been validated to be efficient in
+    // Godbolt.
+    pub(crate) fn cmp(&self, b: &Self) -> std::cmp::Ordering {
+        let xya = ((self.y as u32) << 16) + (self.x as u32);
+        let xyb = ((b.y as u32) << 16) + (b.x as u32);
+        xya.cmp(&xyb)
+    }
+}
+
+// This can be unused when SIMD is selected. Probably a good idea to make it
+// selectable at runtime; will be needed for AVX2.
+#[allow(unused, reason = "may be unused when SIMD is selected")]
+pub(crate) fn render_strips_scalar(
+    tiles: &[Tile],
+    strip_buf: &mut Vec<Strip>,
+    alpha_buf: &mut Vec<u32>,
+) {
+    strip_buf.clear();
+    let mut strip_start = true;
+    let mut cols = alpha_buf.len() as u32;
+    let mut prev_tile = &tiles[0];
+    let mut fp = prev_tile.footprint().0;
+    let mut seg_start = 0;
+    let mut delta = 0;
+    // Note: the input should contain a sentinel tile, to avoid having
+    // logic here to process the final strip.
+    for i in 1..tiles.len() {
+        let tile = &tiles[i];
+        if prev_tile.loc() != tile.loc() {
+            let start_delta = delta;
+            let same_strip = prev_tile.loc().same_strip(&tile.loc());
+            if same_strip {
+                fp |= 8;
+            }
+            let x0 = fp.trailing_zeros();
+            let x1 = 32 - fp.leading_zeros();
+            let mut areas = [[start_delta as f32; 4]; 4];
+            for this_tile in &tiles[seg_start..i] {
+                delta += this_tile.delta();
+                let p0 = Vec2::unpack(this_tile.p0);
+                let p1 = Vec2::unpack(this_tile.p1);
+                let slope = (p1.x - p0.x) / (p1.y - p0.y);
+                for x in x0..x1 {
+                    let startx = p0.x - x as f32;
+                    for y in 0..4 {
+                        let starty = p0.y - y as f32;
+                        let y0 = starty.clamp(0.0, 1.0);
+                        let y1 = (p1.y - y as f32).clamp(0.0, 1.0);
+                        let dy = y0 - y1;
+                        // Note: getting rid of this predicate might help with
+                        // auto-vectorization. That said, just getting rid of
+                        // it causes artifacts (which may be divide by zero).
+                        if dy != 0.0 {
+                            let xx0 = startx + (y0 - starty) * slope;
+                            let xx1 = startx + (y1 - starty) * slope;
+                            let xmin0 = xx0.min(xx1);
+                            let xmax = xx0.max(xx1);
+                            let xmin = xmin0.min(1.0) - 1e-6;
+                            let b = xmax.min(1.0);
+                            let c = b.max(0.0);
+                            let d = xmin.max(0.0);
+                            let a = (b + 0.5 * (d * d - c * c) - xmin) / (xmax - xmin);
+                            areas[x as usize][y] += a * dy;
```

Comment:
I get a panic on this line:
`index out of bounds: the len is 4 but the index is 4`, just running:
```sh
> cargo run --example gpu
```